lib/internal/streams/from.js

```diff
@@ -34,24 +34,53 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitly closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = async function(error, cb) {
+    try {
+      if (needToClose) {
+        needToClose = false;
+        if (typeof iterator.return === 'function') {
+          const { value } = await iterator.return();
+          await value;
+        }
+      }
+    } catch (e) {
+      if (!error) { // Do not hide current error in case if .return() throws
+        error = e;
+      }
+    } finally {
+      cb(error);
+    }
+  };
+
   async function next() {
     try {
+      needToClose = false;
       const { value, done } = await iterator.next();
-      if (done) {
-        readable.push(null);
-      } else if (readable.push(await value)) {
-        next();
-      } else {
-        reading = false;
+      needToClose = !done;
+      const resolved = await value;
+      if (!readable.destroyed) {
+        if (done) {
+          readable.push(null);
+        } else if (readable.push(resolved)) {
+          next();
+        } else {
+          reading = false;
+        }
       }
```

Comment:
The iterator won't be closed in the `else` case here.

Reply:
That's what I meant by checking `readable.destroyed`. `push()`:ing after `destroyed` is safe so you don't need to guard that.

Reply:
Right. Tests are green even without that check.